### PR TITLE
Add Cortex-M7 Architecture and i.MX RT 1052 EVKB Board

### PR DESCRIPTION
**Pull request by valexandru**
### Pull Request Overview

This pull request adds support for the NXP i.MX RT 1052 Crossover MCU and i.MX RT 1052 EVKB Board.

Supported features so far are:

UART
GPIO
Leds
I2C
Timer

This also adds support for the Cortex M7 MCU.

This also adds the fxos8700 component.

### Testing Strategy

This pull request was tested using the i.MX RT 1052 EVKB Board.

### TODO or Help Wanted

This pull request still uses a binary FCB Header as header creation is still a problem. Looking into using https://github.com/imxrt-rs/imxrt-boot-gen . Would this be a problem with the licensing?

Added clock frequency Freq2475MHz for the General Purpose Timer.

### Documentation Updated

- [x] Updated the relevant files in `/docs`.

### Formatting

- [x] Ran `make prepush`.

### New Platform Checklist

- [x] Hardware is widely available.
- [x] I can support the platform, at least initially.
- Basic features are implemented:
  - [x] `Console`, including `debug!()` and userspace `printf()`.
  - [x] Timers.
  - [ ] GPIO with interrupts.

### CODE REVIEW

**Comment by hudson-ayers** [comment]
I don't think we want to make the scheduler timeslice configuration a kernel config option if we can avoid it. In general, I don't understand why the high speed of the MCU matters for those values anyway, as the configuration is specified in microseconds, and your systick implementation should convert that time value to ticks internally, taking into account the MCU frequency.

What issues were you encountering that made you decide to increase those values?

**Comment by lschuermann** [comment]
Thanks for applying the changes of #1913. For the future, please try to avoid opening multiple PRs in order to have a consistent review and discussion history. You should be able to easily change source and destination branch in the GitHub interface, and rebasing or force pushing your branch is always an option.

**Comment by valexandru** [comment]
> I don't think we want to make the scheduler timeslice configuration a kernel config option if we can avoid it. In general, I don't understand why the high speed of the MCU matters for those values anyway, as the configuration is specified in microseconds, and your systick implementation should convert that time value to ticks internally, taking into account the MCU frequency.
> 
> What issues were you encountering that made you decide to increase those values?

I figured out the issue. Even though syst_calib.SKEW = 0 and syst_calib.NOREF = 0, the calibration value
was not corelated to the actual speed of the CPU. If I choose an external clock source, which selects
a 24 MHz oscilator or manually set the frequency according the CPU frequency the systick work correctly.
I created a PR here: [#1987](https://github.com/tock/tock/pull/1987) for this issue. Sorry for the late answer.

**Comment by valexandru** [comment]
> Thanks for applying the changes of #1913. For the future, please try to avoid opening multiple PRs in order to have a consistent review and discussion history. You should be able to easily change source and destination branch in the GitHub interface, and rebasing or force pushing your branch is always an option.
Ok, I will remember next time. Thank you.

File: boards/components/src/fxos8700.rs
Comment (hudson-ayers) [inline boards/components/src/fxos8700.rs:20]: due to how generic components are included, there should be no need for the `#![allow(dead_code)]` that we used in the `imix_components`.

Also, I think it should be better to just remove the unused_import than to allow it.

File: chips/imxrt1050/src/lib.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/lib.rs:7]: As part of the effort to move to stable Rust, we have removed all uses of the in_band_lifetimes feature in Tock. Can you remove it from this PR please?

File: doc/Startup.md
Comment (hudson-ayers) [inline doc/Startup.md:102]: good catch!

**Review by hudson-ayers — DISMISSED** [review]
Just a couple comments from an incomplete pass

File: doc/Startup.md
Comment (valexandru) [reply doc/Startup.md]: Oh, sorry about that :))

File: arch/cortex-m7/src/mpu.rs
Comment (bradjc) [inline arch/cortex-m7/src/mpu.rs:1]: Can this just be a symlink to the m3 mpu file?

File: boards/README.md
Comment (bradjc) [inline boards/README.md:23]:
```suggestion
| [Digilent Arty A-7 100T](arty_e21/README.md)                         | RISC-V RV32IMAC | SiFive E21     | openocd    | tockloader     |
```

File: boards/imxrt1050-evkb/chip_layout.ld
Comment (bradjc) [inline boards/imxrt1050-evkb/chip_layout.ld:1]: update

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (bradjc) [inline boards/imxrt1050-evkb/src/boot_header.rs:1]: Can you comment where this comes from?

Comment (bradjc) [reply boards/imxrt1050-evkb/src/boot_header.rs:1]: Do we need the ` boards/imxrt1050-evkb/boot_header` file too?

File: boards/imxrt1050-evkb/src/main.rs
Comment (bradjc) [inline boards/imxrt1050-evkb/src/main.rs:7]:
```suggestion
```

File: boards/imxrt1050-evkb/README.md
Comment (bradjc) [inline boards/imxrt1050-evkb/README.md:4]: This takes me to a chip page. Can you link to the board as well?

File: boards/imxrt1050-evkb/src/main.rs
Comment (bradjc) [inline boards/imxrt1050-evkb/src/main.rs:418]: You might consider updating this with the change in #1962.

File: chips/imxrt1050/README.md
Comment (bradjc) [inline chips/imxrt1050/README.md:1]:
```suggestion
Imxrt1050 Series MCU
====================
```

File: chips/imxrt1050/src/chip.rs
Comment (bradjc) [inline chips/imxrt1050/src/chip.rs:65]:
```suggestion
```

File: arch/cortex-m7/src/mpu.rs
Comment (valexandru) [reply arch/cortex-m7/src/mpu.rs:1]: Cortex-M7 has 16 MPU regions, while Cortex-M3 has only 8 regions

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (valexandru) [reply boards/imxrt1050-evkb/src/boot_header.rs:1]: No, from boot_header was the vector extracted. I will delete the file now.

File: boards/imxrt1050-evkb/README.md
Comment (alistair23) [inline boards/imxrt1050-evkb/README.md:65]: This can't be the only way to flash things to the board? There is no command line tool?

**Comment by bradjc** [comment]
Marking as blocked just because we have to reorganize the arch/cortex-m* crates before adding another copy...it's getting to be unmaintainable.

**Comment by valexandru** [comment]
@bradjc Can i help with something in the process?

**Comment by bradjc** [comment]
You can try using #2027 and seeing if that works for you, or suggesting any updates to #2027. That should substantially help reduce the duplicated code.

**Comment by hudson-ayers** [comment]
Now that #2027 is merged, this is no longer blocked. It needs to be rebased onto current master such that it does not re-duplicate all of the cortex-m code, however, and only cortex-m7 specific code should be in its own folder in `arch/`.

**Comment by valexandru** [comment]
> Now that #2027 is merged, this is no longer blocked. It needs to be rebased onto current master such that it does not re-duplicate all of the cortex-m code, however, and only cortex-m7 specific code should be in its own folder in `arch/`.

I rebased my branch. I think everything should be ok now.

File: boards/imxrt1050-evkb/Cargo.toml
Comment (hudson-ayers) [inline boards/imxrt1050-evkb/Cargo.toml:16]: Why is this build dependency needed? Generally, Tock has a policy against external dependencies, which I think applies to build dependencies as well.

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (hudson-ayers) [reply boards/imxrt1050-evkb/src/boot_header.rs:1]: I am somewhat confused by this conversation, should this file have been deleted?

File: chips/imxrt1050/src/deferred_call_tasks.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/deferred_call_tasks.rs:30]: If there are no deferred calls in use for this chip yet, I am not sure we should include this file or the other machinery in `chip.rs` for deferred call tasks (unless you are using deferred calls out of tree and plan to add those in subsequent PRs).

File: chips/imxrt1050/src/lib.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/lib.rs:9]: Can you remove this allow and resolve the warnings it is hiding?

File: doc/courses/rustconf/presentation/slides.md
Comment (hudson-ayers) [inline doc/courses/rustconf/presentation/slides.md:122]: haha, I realize this was probably the result of an automated change, but lets leave the old presentations as they were :)

File: chips/imxrt1050/src/lpuart.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/lpuart.rs:301]:
```suggestion
```

File: chips/imxrt1050/src/lpuart.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/lpuart.rs:309]:
```suggestion
```

File: chips/imxrt1050/src/lpi2c.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/lpi2c.rs:780]: I suspect these `debug!()` statements be removed before merging

File: chips/imxrt1050/src/gpio.rs
Comment (hudson-ayers) [inline chips/imxrt1050/src/gpio.rs:1110]: The porting guide (https://github.com/tock/tock/blob/master/doc/Porting.md) specifies that GPIO with interrupts is a prerequisite for a chip to be added to Tock

File: boards/imxrt1050-evkb/src/main.rs
Comment (hudson-ayers) [inline boards/imxrt1050-evkb/src/main.rs:62]: In https://github.com/tock/tock/pull/2001 we moved to a design that no longer relies on this Hack, could you incorporate that approach here as well?

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
I did another pass, mostly minor comments. The important ones are:

1) can we get rid of the `cc` build dependency?

2) Use a symlink for the cortex-m7 MPU instead of copying the file.

File: arch/cortex-m7/src/mpu.rs
Comment (valexandru) [reply arch/cortex-m7/src/mpu.rs]: The difference between them is the number of mpu regions used (16 for Cortex-M7 instead of just 8). For some weird reasons the file was not updated into the repository.

File: boards/imxrt1050-evkb/Cargo.toml
Comment (valexandru) [reply boards/imxrt1050-evkb/Cargo.toml]: Oh, i used it in the beginning when i compiled the boot header. Now it is no longer needed, so I deleted it.

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (valexandru) [reply boards/imxrt1050-evkb/src/boot_header.rs:1]: There was a binary file with a similar name that I deleted before. This is the file used now and included the main.rs of the board for the boot header.

File: chips/imxrt1050/src/deferred_call_tasks.rs
Comment (valexandru) [reply chips/imxrt1050/src/deferred_call_tasks.rs]: I do not use them, so I deleted it.

File: chips/imxrt1050/src/lib.rs
Comment (valexandru) [reply chips/imxrt1050/src/lib.rs]: Done.

File: doc/courses/rustconf/presentation/slides.md
Comment (valexandru) [reply doc/courses/rustconf/presentation/slides.md]: Oh, sorry about that :)) .  There shouldn't be any other left.

File: chips/imxrt1050/src/lpi2c.rs
Comment (valexandru) [reply chips/imxrt1050/src/lpi2c.rs]: Deleted them.

File: boards/imxrt1050-evkb/src/main.rs
Comment (valexandru) [reply boards/imxrt1050-evkb/src/main.rs]: Done.

**Comment by hudson-ayers** [comment]
I think this is now just blocked on GPIO interrupt support. The small MPU differences requiring a duplicate file is unfortunate, but can't easily be fixed without const generics, as I don't think the macro approach we use for RISC-V is worthwhile here.

File: boards/imxrt1050-evkb/chip_layout.ld
Comment (bradjc) [inline boards/imxrt1050-evkb/chip_layout.ld:11]: Does this work? I would expect a lot of MPU errors with large-ish apps because not many powers of two application sizes are aligned here. I think it would be better to make the `rom` 0x2000 smaller and start the `prog` at 0x63000000.

**Review by bradjc — APPROVED** [review]

**Comment by valexandru** [comment]
> I think this is now just blocked on GPIO interrupt support. The small MPU differences requiring a duplicate file is unfortunate, but can't easily be fixed without const generics, as I don't think the macro approach we use for RISC-V is worthwhile here.

Ok. I am working on the GPIO interrupt support now

File: boards/imxrt1050-evkb/chip_layout.ld
Comment (valexandru) [reply boards/imxrt1050-evkb/chip_layout.ld:11]: So far, I had no issues with this setup, but I can see your point. I will update it.

Comment (valexandru) [reply boards/imxrt1050-evkb/chip_layout.ld]: It is updated now, along with the GPIO interrupt support.

**Comment by valexandru** [comment]
For some reason, the last two tasks: ci-qemu and deploy-preview are failing. For deploy preview, I can not even see the details. It just tells me: "You might not have permissions to see this page."

**Comment by hudson-ayers** [comment]
> For some reason, the last two tasks: ci-qemu and deploy-preview are failing. For deploy preview, I can not even see the details. It just tells me: "You might not have permissions to see this page."

We have been seeing spurious errors for both checks lately, don't worry about it for now. qemu.org being down is one of the issues

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (bradjc) [inline boards/imxrt1050-evkb/src/boot_header.rs:4]: Isn't it possible to define the section of this directly in rust, without needing an assembly file?

File: chips/imxrt1050/README.md
Comment (bradjc) [inline chips/imxrt1050/README.md:1]:
```suggestion
Imxrt1050 Series MCU
================
```

**Review by bradjc — APPROVED** [review]

**Review by hudson-ayers — APPROVED** [review]
This looks good to me. The new chip/board will eventually need to be updated to use the new peripheral instantiation approach (#2069) but I think it can be merged as-is and switched over in a separate PR.

File: boards/imxrt1050-evkb/src/boot_header.rs
Comment (valexandru) [reply boards/imxrt1050-evkb/src/boot_header.rs:4]: For the moment that is the only solution that I managed to implement. I am working with someone from NXP in order to see how to write to code for it, however it is really complicated and will take a lot of time to write.

**Comment by valexandru** [comment]
> This looks good to me. The new chip/board will eventually need to be updated to use the new peripheral instantiation approach (#2069) but I think it can be merged as-is and switched over in a separate PR.

Ok, perfect.

**Review by bradjc — APPROVED** [review]

**Pull request merged**
